src/sdk/namespace_cache.js

```diff
@@ -41,43 +54,104 @@ class NamespaceCache {
     /////////////////
 
     async read_object_md(params, object_sdk) {
+
+        let object_info_cache = null;
+        let cache_etag = '';
         try {
-            const object_info = await this.namespace_nb.read_object_md(params, object_sdk);
+            const get_from_cache = params.get_from_cache;
+            // Remove get_from_cache if exists for maching RPC schema
+            params = _.omit(params, 'get_from_cache');
+            object_info_cache = await this.namespace_nb.read_object_md(params, object_sdk);
+            if (get_from_cache) {
+                dbg.log0('NamespaceCache.read_object_md get_from_cache is enabled', object_info_cache);
+                object_info_cache.should_read_from_cache = true;
+                return object_info_cache;
+            }
 
-            // TODO this is the wrong condition - object create_time should be preserved 
-            // from hub so not the validation time, so we need a custom cache metadata for that.
-            const cache_validation_time = object_info.create_time;
+            const cache_validation_time = object_info_cache.cache_last_valid_time;
             const time_since_validation = Date.now() - cache_validation_time;
 
-            const CACHE_TTL = 60000; // 1 minute
-            if (time_since_validation <= CACHE_TTL) {
-                object_info.should_read_from_cache = true; // mark it for read_object_stream
-                console.log('NamespaceCache.read_object_md: from cache', object_info);
-                return object_info;
+            if ((this.caching.ttl_ms > 0 && time_since_validation <= this.caching.ttl_ms) || this.caching.ttl_ms < 0) {
+                object_info_cache.should_read_from_cache = true; // mark it for read_object_stream
+                dbg.log0('NamespaceCache.read_object_md use md from cache', object_info_cache);
+                return object_info_cache;
             }
 
+            cache_etag = object_info_cache.etag;
         } catch (err) {
-            console.warn('NamespaceCache.read_object_md: cache error', err);
+            dbg.log0('NamespaceCache.read_object_md: error in cache', err);
         }
 
-        // TODO do we have any benefit in caching just object mds?
-        return this.namespace_hub.read_object_md(params, object_sdk);
+        let object_info_hub = null;
+        try {
+            object_info_hub = await this.namespace_hub.read_object_md(params, object_sdk);
+            if (object_info_hub.etag === cache_etag) {
+                dbg.log0('NamespaceCache.read_object_md: same etags: updating cache valid time', object_info_hub);
+                process.nextTick(() => {
```

Comment:
Same comments as before - use an internal async method on the namespace class such as `async _update_cache_last_valid_time()` which will do its internal try-catch-await and then call it from here with `setImmediate(() => this.__update_cache_last_valid_time(...))`

Reply:
ok